File: VB/GridControlCRUD.Common/ViewModel - VB/VirtualCollectionViewModel.vb
Imports DevExpress.CRUD.DataModel
Imports DevExpress.Mvvm
Imports DevExpress.Mvvm.DataAnnotations
Imports DevExpress.Xpf.Data
Imports DevExpress.Mvvm.Xpf
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Linq.Expressions
Imports System.Threading.Tasks
Imports System.ComponentModel
Imports DevExpress.Xpf.Core

Namespace DevExpress.CRUD.ViewModel
	Public MustInherit Class VirtualCollectionViewModel(Of T As {Class, New})
		Inherits ViewModelBase

		Private ReadOnly dataProvider As IDataProvider(Of T)

		Protected Sub New(ByVal dataProvider As IDataProvider(Of T))
			Me.dataProvider = dataProvider
			StartRefresh()
		End Sub

		Public ReadOnly Property FilterType() As Type
			Get
				Return GetType(Expression(Of Func(Of T, Boolean)))
			End Get
		End Property
		Private ReadOnly Property DialogService() As IDialogService
			Get
				Return GetService(Of IDialogService)()
			End Get
		End Property

		<Command>
		Public Sub Fetch(ByVal args As FetchRowsAsyncArgs)
			args.Result = dataProvider.GetQueryableResultAsync(Of T, FetchRowsResult)(Function(queryable)
				Return queryable.SortBy(args.SortOrder, defaultUniqueSortPropertyName:= dataProvider.KeyProperty).Where(CType(args.Filter, Expression(Of Func(Of T, Boolean)))).Skip(args.Skip).Take(If(args.Take, 30)).ToArray()
			End Function)
		End Sub

		<Command>
		Public Sub GetTotalSummaries(ByVal args As GetSummariesAsyncArgs)
			args.Result = dataProvider.GetQueryableResultAsync(Function(queryable)
				Return queryable.Where(CType(args.Filter, Expression(Of Func(Of T, Boolean)))).GetSummaries(args.Summaries)
			End Function)
		End Sub

		<Command>
		Public Sub GetUniqueValues(ByVal args As GetUniqueValuesAsyncArgs)
			args.ResultWithCounts = dataProvider.GetQueryableResultAsync(Function(queryable)
				Return queryable.Where(CType(args.Filter, Expression(Of Func(Of T, Boolean)))).DistinctWithCounts(args.PropertyName)
			End Function)
		End Sub

		<Command>
		Public Sub OnUpdate(ByVal args As EntityUpdateArgs)
			Dim entity = DirectCast(args.Entity, T)
			Dim commands = CreateCommands(Sub()
				dataProvider.Update(entity)
				args.Updated = True
			End Sub)
			DialogService.ShowDialog(commands, "Edit " & GetType(T).Name, CreateEntityViewModel(entity))
		End Sub

		<Command>
		Public Sub OnCreate(ByVal args As EntityCreateArgs)
			Dim entity = New T()
			Dim commands = CreateCommands(Sub()
				dataProvider.Create(entity)
				args.Entity = entity
			End Sub)
			DialogService.ShowDialog(commands, "New " & GetType(T).Name, CreateEntityViewModel(entity))
		End Sub

		Private Function CreateCommands(ByVal saveAction As Action) As UICommand()
			Return { New UICommand(Nothing, "Save", New DelegateCommand(Of CancelEventArgs)(Function(cancelArgs)
				Try
					saveAction()
				Catch e As Exception
					GetService(Of IMessageBoxService)().ShowMessage(e.Message, "Error", MessageButton.OK)
					cancelArgs.Cancel = True
				End Try
			End Function), isDefault:= True, isCancel:= False), New UICommand(Nothing, "Cancel", Nothing, isDefault:= False, isCancel:= True)}
		End Function

		Protected MustOverride Function CreateEntityViewModel(ByVal entity As T) As EntityViewModel(Of T)

		<Command>
		Public Sub OnDelete(ByVal args As RowDeleteArgs)
			Me.dataProvider.Delete(DirectCast(args.Row, T))
		End Sub

		<Command>
		Public Sub OnRefresh(ByVal args As RefreshArgs)
			args.Result = OnRefreshCoreAsync()
		End Sub
		Private Async Sub StartRefresh()
			Await OnRefreshCoreAsync()
		End Sub
		Protected Overridable Function OnRefreshCoreAsync() As Task
			Return Task.CompletedTask
		End Function
	End Class
	Public Class EntityViewModel(Of T)
		Inherits ViewModelBase

		Public Sub New(ByVal entity As T)
			Me.Entity = entity
		End Sub
		Public ReadOnly Property Entity() As T
	End Class
End Namespace

File: CS/GridControlCRUD.Common/ViewModel/VirtualCollectionViewModel.cs
using DevExpress.CRUD.DataModel;
using DevExpress.Mvvm;
using DevExpress.Mvvm.DataAnnotations;
using DevExpress.Xpf.Data;
using DevExpress.Mvvm.Xpf;
using System;
using System.Linq;
using System.Linq.Expressions;
using System.Threading.Tasks;

namespace DevExpress.CRUD.ViewModel {
    public abstract class VirtualCollectionViewModel<T> : ViewModelBase where T : class, new() {
        readonly IDataProvider<T> dataProvider;

        public IMessageBoxService MessageBoxService { get { return GetService<IMessageBoxService>(); } }

        protected VirtualCollectionViewModel(IDataProvider<T> dataProvider) {
            this.dataProvider = dataProvider;
            StartRefresh();
        }

        public Type FilterType => typeof(Expression<Func<T, bool>>);

        [Command]
        public void Fetch(FetchRowsAsyncArgs args) {
            args.Result = dataProvider.GetQueryableResultAsync<T, FetchRowsResult>(queryable => {
                return queryable
                    .SortBy(args.SortOrder, defaultUniqueSortPropertyName: dataProvider.KeyProperty)
                    .Where((Expression<Func<T, bool>>)args.Filter)
                    .Skip(args.Skip)
                    .Take(args.Take ?? 30)
                    .ToArray<object>();
            });
        }

        [Command]
        public void GetTotalSummaries(GetSummariesAsyncArgs args) {
            args.Result = dataProvider.GetQueryableResultAsync(queryable => {
                return queryable
                    .Where((Expression<Func<T, bool>>)args.Filter)
                    .GetSummaries(args.Summaries);
            });
        }

        [Command]
        public void GetUniqueValues(GetUniqueValuesAsyncArgs args) {
            args.ResultWithCounts = dataProvider.GetQueryableResultAsync(queryable => {
                return queryable
                    .Where((Expression<Func<T, bool>>)args.Filter)
                    .DistinctWithCounts(args.PropertyName);
            });
        }

        [Command]
        public void OnCreateRow(RowValidationArgs args) {
            if(args.IsNewItem) {
                dataProvider.Create((T)args.Item);
            } else {
                dataProvider.Update((T)args.Item);
            }
        }

        [Command]
        public void OnDeleteRow(DeleteRowsValidationArgs args) {
            var row = (T)args.Items[0];
            if(row == null)
                return;
            if(MessageBoxService.ShowMessage("Are you sure you want to delete this row?", "Delete Row", MessageButton.OKCancel) == MessageResult.Cancel) {
                args.Result = "Canceled";
            }
            this.dataProvider.Delete(row);
        }

        [Command]
        public void OnRefresh(RefreshArgs args) {
            args.Result = OnRefreshCoreAsync();
        }
        async void StartRefresh() {
            await OnRefreshCoreAsync();
        }
        protected virtual Task OnRefreshCoreAsync() {
            return Task.CompletedTask;
        }
    }
    public class EntityViewModel<T> : ViewModelBase {
        public EntityViewModel(T entity) {
            Entity = entity;
        }
        public T Entity { get; }
    }
}
